### PR TITLE
Trigger automation action

### PR DESCRIPTION
**Pull request by mjashanks**
## Description

#949 

Changes
1. When creating a new automation, we now have an "App Action" trigger. This allows the manual building of the input schema.
2. For button click "Define Actions", we have a new option, called "Trigger Automation", which allows selecting of automations that have the "App Action" trigger type. We can then data-bind, to build up the schema declared in No. 1 above
3. The client library handles this new "Trigger Automation" type.

## Screenshots
https://www.loom.com/share/a8abec4b22a249eb937df2fbc3613e8b

### CODE REVIEW

**Comment by joebudi** [comment]
@mjashanks the input style for the automation item is not consistent with the rest of the inputs. (It's white, the others are grey).

**Comment by mjashanks** [comment]
> @mjashanks the input style for the automation item is not consistent with the rest of the inputs. (It's white, the others are grey).

@joebudi do you mean the outlined one, where we type into to add fields? Will update that.

image

How do you feel about these?

image

I realise that they are very different from other inputs. We are short on space here, which is why I opted for a plain, squared-off input.

File: packages/server/src/automations/triggers.js [hunk 1]
```diff
@@ -119,6 +120,37 @@ const BUILTIN_DEFINITIONS = {
     },
     type: "TRIGGER",
   },
+  APP: {
```
Comment (shogunpurple) [inline packages/server/src/automations/triggers.js:123]: should we just call this `AUTOMATION_TRIGGER`? App action seems quite generic

File: packages/builder/src/components/userInterface/EventsEditor/actions/TriggerAutomation.svelte [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+    return {
+      name: automation.name,
+      _id: automation._id,
+      schema,
+    }
+  }
+
+  $: automations = $automationStore.automations
+    .filter(a => a.definition.trigger && a.definition.trigger.stepId === "APP")
```
Comment (shogunpurple) [inline packages/builder/src/components/userInterface/EventsEditor/actions/TriggerAutomation.svelte:21]: NAB: should be able to just do: `a.definition.trigger?.stepId === "APP"`

File: packages/client/src/utils/buttonActions.js [hunk 3]
```diff
@@ -21,6 +21,17 @@ const deleteRowHandler = async (action, context) => {
   })
 }
 
+const triggerAutomationHandler = async (action, context) => {
+  const params = {}
+  for (let field in action.parameters.fields) {
```
Comment (shogunpurple) [inline packages/client/src/utils/buttonActions.js:26]: I think you might be able use the `enrichDataBindings` helper here to parse all your values without looping.

File: packages/server/src/automations/triggers.js [hunk 4]
```diff
@@ -200,12 +232,19 @@ async function fillRowOutput(automation, params) {
...
-    automation.definition.trigger != null &&
-    automation.definition.trigger.inputs.tableId != null
-  ) {
-    params = await fillRowOutput(automation, params)
+  if (automation.definition != null && automation.definition.trigger != null) {
+    if (automation.definition.trigger.inputs.tableId != null) {
+      params = await fillRowOutput(automation, params)
+    }
+    if (automation.definition.trigger.stepId === "APP") {
+      // values are likely to be submitted as strings, so we shall convert to correct type
```
Comment (shogunpurple) [inline packages/server/src/automations/triggers.js:240]: Should we just convert them before sending?

File: packages/builder/src/components/automation/SetupPanel/SchemaSetup.svelte [hunk 5]
```diff
@@ -0,0 +1,113 @@
+<script>
+  import { Input } from "@budibase/bbui"
+
+  export let value = {}
+  $: fieldsArray = Object.entries(value).map(([name, type]) => ({
```
Comment (shogunpurple) [inline packages/builder/src/components/automation/SetupPanel/SchemaSetup.svelte:5]: NAB: I think you might save yourself a lot of pain here if you just dealt with the array the whole time, used `bind`, and construct/reduce the schema object from the array when the actual action gets executed.

**Review by shogunpurple — CHANGES_REQUESTED** [review]
Mainly stylistic things, but I think we need to rethink the design/styling of the schema builder inputs before we merge this.

**Comment by mjashanks** [comment]
> I think you might be able use the enrichDataBindings helper here to parse all your values without looping.

i cannot, as the fields object is a `{ "key" : { ... } }`. enrichBindings expects a `{ "key": "string" }`

> Should we just convert them before sending?

The issue with this is that we need to then know the type on the client app, and the type can be changed in trigger setup. At the moment, we need to be able to change the type, as all fields default to `string`, when creating an automation via a button click.

> I think you might save yourself a lot of pain here if you just dealt with the array the whole time, used bind, and construct/reduce the schema object from the array when the actual action gets executed.

I think you're right. However, I am reusing the `SaveFields` control, which is also used by the `Save Row` action. I don't want to have to go changing that as well.

Other changes adressed

**Review by shogunpurple — APPROVED** [review]

**Comment by mjashanks** [comment]
Demo of creating an automation from button actions, and some styling changes to the schema setup, designed by @joebudi:

https://www.loom.com/share/2d4175c0e1bd4659bb51f7cf8cfe8d60

**Comment by joebudi** [comment]
> Demo of creating an automation from button actions, and some styling changes to the schema setup, designed by @joebudi:
> 
> 
> https://www.loom.com/share/2d4175c0e1bd4659bb51f7cf8cfe8d60

I was semi-expecting the add button to open a popover. I'm not sure how the user would know to type the title of the new field or dropdown the string - I feel these need indicators. What do you think?

**Comment by mjashanks** [comment]
> I was semi-expecting the add button to open a popover.

Yes, it would be better, although I think the user would work out what they are supposed to do.  

How would you expect the user to edit an existing field? delete and re-add? At the moment, they can just into the fields.

Happy to add a popover, although I will not be getting back to this until Friday at the earliest.

**Comment by joebudi** [comment]
> > I was semi-expecting the add button to open a popover.
> 
> Yes, it would be better, although I think the user would work out what they are supposed to do.
> 
> How would you expect the user to edit an existing field? delete and re-add? At the moment, they can type into the fields.
> 
> Happy to add a popover, although I will not be getting back to this until Friday at the earliest.

Can you add a title placeholder? and maybe a dropdown arrow? That would go a long way IMO.

**Comment by mjashanks** [comment]
@joebudi hows this?

image

**Comment by joebudi** [comment]
> @joebudi hows this?
> 
> image

Bingo. Incredible work Mike. What does it look like in dark mode?

**Comment by mjashanks** [comment]
image

**Comment by joebudi** [comment]
Chef's kiss

**Comment by mjashanks** [comment]
@shogunpurple there have been a few changes since you reviewed. Let me know if you are still happy :)

**Comment by shogunpurple** [comment]
yep! LGTM

**Pull request merged**